Rewrite coordsCharInner

marijnh opened the pull request:
Hi @adrianheine, do you have time to take a look at this? Since you were the last to think about this stuff, you might be able to spot mistakes that I didn't notice.

I basically made `coordsCharInner` for bidi lines run two binary searches, one on the bidi spans and one on the actual characters of a single span. It's conceptually simpler, and probably faster (though I didn't benchmark). Since the other algorithm was found to be running into its infinite loop guard in some cases (see issue #4926), I figured it'd be better to start over rather than try to salvage that.

The main question is whether doing a binary search on the bidi spans makes sense on wrapped lines. I think it does, because as we also found with `wrappedLineExtent`, even when an rtl span is wrapped, as a whole it remains inside the computed ordering.

adrianheine commented:
Whoa, I'll look into it. I don't really understand your last sentence, though. With wrapped lines, a single visual line always contains a _logically_ continuous number of chars. Inside a visual line, all bidi spans that are present are __visually__ ordered as in the bidi order. There might be spans missing, though. I was thinking about using the bidi spans instead of `moveVisually` yesterday as well.

adrianheine commented on src/measurement/position_measurement.js:
```diff
@@ -446,66 +452,106 @@ export function wrappedLineExtentChar(cm, lineObj, preparedMeasure, target) {
+                               "line", lineObj, preparedMeasure)
+      if (boxIsAfter(start, x, y, true) && start.top > y) {
+        part = order[--index]
+        ltr = part.level % 2 == 0
```
This is wrong in cases with lvl 2, I think. Example:

http://unicode.org/cldr/utility/bidi.jsp?a=abcd+%D9%85+1+%D8%B1&p=LTR&hack=on

If your line fits 6 characters, according to [UBA](http://unicode.org/reports/tr9/#Reordering_Resolved_Levels), it will be  
```
012345
------
abcd E 
N F
```

But the bidi ordering is:

```
"{"level":0,"from":0,"to":5}, {"level":1,"from":8,"to":10}, {"level":2,"from":7,"to":8}, {"level":1,"from":5,"to":7}"
```

The first bidi part in bidi ordering order that's after, say, the right end of `E` will be `order[1]`. I suppose you are looking for `order[3]` though?

marijnh replied:
The unicode.org link isn't loading right now, so I'm not sure what you mean, but I guess you're saying that, in the presence of wrapped lines with level 2 spans in them, the visual ordering may not correspond to the ordering in `order`? Experiments seem to conform that. That's bad!

I'll see if I can work around this by doing the clipping to a single wrapped line _before_ finding the appropriate bidi part (discarding those outside of the line's range). I guess that makes applying a binary search a lot harder, but I suppose falling back to a full search in the wrapped case (and discarding all spans that aren't in the line) should be okay.

adrianheine replied:
Hm, it's loading for me. Anyway, what I am saying is: There are cases in which a span `A` that comes before a span `B` _in the bidi order_ comes after the span `B` _in top-down order_.

marijnh commented:
Yeah, I see your point, I've added a test case for that and fixed it the way I described (different search for bidi spans when in wrapped content).

The total complexity of this is now at least as bad as the old algorithm (though I added more comments so the line count isn't representative), but I think the new code is a little easier to reason about and hopefully more robust.

marijnh commented:
I've gone ahead and merged this as 68b4f44e3c6e, but if you have any further concerns definitely let me know.
